README.md
# Unity Cloud Build Viewer VSCode extension

This extension is for viewing Unity Cloud Build build logs in much stress-free way.

# Features

* Browse latest builds on `Unity Cloud Build` explorer view.
* Show build details.
* Load build logs progressively.
  * Can be launched from build detail view, or context menu of the `Unity Cloud Build` explorer view.
* Trigger rebuild/cancel from build history.

# Setup

`Api Key` setting in the `Unity Cloud Build Viewer` section must be set before using this plugin. The correct value can be found on the official `Unity Cloud Build` dashboard's `Settings` screen.

After that, push `Ctrl+P` to show the command palette and select `'Unity Cloud Build Viewer: setup'` command.
If the api key was correct, you can choose a target project from existing projects.
Then, `Unity Cloud Build` explorer view will work and it will show the latest builds.

# Development

to publish:

```sh
vsce publish
```

# License

MIT


CHANGELOG.md
# Change Log

* Version 0.0.1
  * initial release

* Version 0.0.2
  * fix setup failure

* Version 0.0.3
  * added 'build again'
  * implemented build filter
  * added branch and commit info on build detail view

* Version 0.0.4
  * added "Reload Log" and "View Text Log" context menu items.
  * show old running builds at the top of the list.

* Version 0.0.5
  * fix paging of builds

* Version 0.0.6
  * fix bug at persisting project values

* Version 0.0.7
  * show more pages for filtered builds
  * add build time / current time since build start on build detail view.

* Version 0.0.8
  * add time display like '3h ago'

* Version 0.0.9
  * add share link on build detail page
  * add Create Share Link command for builds

* Version 0.1.0
  * fix share format on build detail page
  * add canceling of builds

* Version 0.1.1
  * add caching of build targets

* Version 0.1.2
  * show build config on build detail view


src/apiloader.ts
import * as vscode from 'vscode';
import OpenAPIClientAxios from 'openapi-client-axios';
import { Client as CloudBuildClient, Paths } from './cloudbuildapi';
import axios from 'axios'

const cloudBuildLogUrlBase = "https://build-api.cloud.unity3d.com";
const cloudBuildLogScheme = "unitycloudbuildviewer";

type GetBuildsResponseItemType = Paths.GetBuilds.Responses.$200[0];
type GetBuildTargetsResponseItemType = Paths.GetBuildTargets.Responses.$200[0];
type GetListProjectsForUserResponseItemType = Paths.ListProjectsForUser.Responses.$200[0];
type GetShareResponseItemType = Paths.GetShare.Responses.$200;
type GetBuildTargetResponseItemType = Paths.GetBuildTarget.Responses.$200;

class BuildTargetsCache
{
    buildTargets : BuildTargetInfo[]
    reloadTime : number

    currentBuildTargets() : BuildTargetInfo[] {
        return this.isExpired() ? null : this.buildTargets;
    }

    update(buildTargets : BuildTargetInfo[]) {
        this.buildTargets = buildTargets;
        this.reloadTime = Date.now();
    }

    isExpired() : boolean {
        return Date.now() - this.reloadTime > 3 * 24 * 60 * 60 * 1000; // keep cache 3 days
    }
}

export class ApiLoader
{
    context: vscode.ExtensionContext;
    definitionFilePath: string;
	apiKeyProvider: () => string;
    buildTargetsCache = new BuildTargetsCache();

    api: OpenAPIClientAxios;
    currentApiKey: string;

    constructor(apiKeyProvider: () => string, context: vscode.ExtensionContext, definitionFilePath: string) {
        this.definitionFilePath = definitionFilePath;
        this.context = context;
        this.apiKeyProvider = apiKeyProvider;
        this.checkApiKey();
    }

    checkApiKey() {
        const apiKey = this.apiKeyProvider();
        if (apiKey != this.currentApiKey) {
            this.currentApiKey = apiKey;
            this.recreateApi();
        }
    }

    getOrgAndProjectParams() : {[name: string] :string} {
        const orgId: string = this.context.globalState.get("orgId");
        const projectId: string = this.context.globalState.get("projectId");
        console.log("orgId: " + orgId + " projectId: " + projectId);
        if (orgId == null || projectId == null) {
            throw new Error("UnityCloudBuild Viewer Not configured correctly. Please run setup.");
        }
        return {orgid: orgId, projectid: projectId};
    }

    recreateApi() {
        if (this.currentApiKey == "") {
            this.api = null;
            return;
        }
        this.api = new OpenAPIClientAxios(
            {
                definition: this.definitionFilePath,
                axiosConfigDefaults: {
                    headers: {
                        'Content-Type': 'application/json',
                        'Authorization': 'Basic ' + this.currentApiKey,
                    },
                },
                withServer: { url: 'https://build-api.cloud.unity3d.com/api/v1/', description: 'cloud build api server' }
            }
        );
    }

    isConfigured(): boolean {
        this.checkApiKey();
        if (this.currentApiKey == "") {
            return false;
        }
        const orgId: string = this.context.globalState.get("orgId");
        const projectId: string = this.context.globalState.get("projectId");
        if (orgId == null || projectId == null) {
            return false;
        }
        return true;
    }

    async getBuilds(buildTargetId?: string, buildStatus?: string, page?: number, perPage?: number): Promise<BuildInfo[]>
    {
        this.checkApiKey();
        const orgAndProject = this.getOrgAndProjectParams();
        const client = await this.api.init<CloudBuildClient>();
        const res = await client.getBuilds({...orgAndProject, 
            buildtargetid: (buildTargetId ?? "_all"), 
            buildStatus: buildStatus,
            page: page ?? 1,
            per_page: perPage ?? 30});
        const builds = res.data;
        const buildInfos = builds.map(i => new BuildInfo(i));
        return buildInfos;
    }

    async readProjects() : Promise<ProjectInfo[]> {
        this.checkApiKey();
        const client = await this.api.init<CloudBuildClient>();
        const res = await client.listProjectsForUser();
        console.log('projects', res.data);
        return res.data.map(p => new ProjectInfo(p));
    }

    async startBuild(buildTargetId: string) {
        this.checkApiKey();
        const orgAndProject = this.getOrgAndProjectParams();
        const client = await this.api.init<CloudBuildClient>();
        const res = await client.startBuilds({...orgAndProject, buildtargetid: buildTargetId})
        console.log('startBuild result:', res.data);
        console.log('startBuild status:', res.data[0].buildStatus);
        return res.data[0];
    }

    async getBuildTargets() : Promise<BuildTargetInfo[]> {
        this.checkApiKey();
        const orgAndProject = this.getOrgAndProjectParams();
        const client = await this.api.init<CloudBuildClient>();
        const res = await client.getBuildTargets({...orgAndProject})
        console.log('getBuildTargets result:', res.data);
        return res.data.map(p => new BuildTargetInfo(p));
    }

    async getBuildTargetsCached(reload: boolean) : Promise<BuildTargetInfo[]> {
        const cachedBuildTargets = this.buildTargetsCache.currentBuildTargets();
        if (!reload && cachedBuildTargets != null) {
            return cachedBuildTargets;
        }
        var buildTargets = await this.getBuildTargets();
        if (buildTargets != null) {
            this.buildTargetsCache.update(buildTargets);
        }
        return buildTargets;
    }

    async cancelBuild(buildTargetId: string, buildNumber: number) : Promise<string> {
        this.checkApiKey();
        const orgAndProject = this.getOrgAndProjectParams();
        const client = await this.api.init<CloudBuildClient>();
        const res = await client.cancelBuild({...orgAndProject, buildtargetid: buildTargetId, number: buildNumber})
        console.log('cancelBuild result:', res.data);
        return res.data;
    }

    async getShare(buildTargetId: string, buildNumber: number) : Promise<ShareInfo> {
        this.checkApiKey();
        const orgAndProject = this.getOrgAndProjectParams();
        const client = await this.api.init<CloudBuildClient>();
        const res = await client.getShare({...orgAndProject, buildtargetid: buildTargetId, number: buildNumber})
        console.log('getShare result:', res.data);
        const share = res.data as GetShareResponseItemType;
        if (share == null) {
            throw "getShare failed.";
        }
        return new ShareInfo(share);
    }

    // for create share, url is provided inside build info.
    async createShare(url: string) {
        this.checkApiKey();
        const orgAndProject = this.getOrgAndProjectParams();
        const res = await axios.post(url, {
            headers: {
                'Content-Type': 'application/json',
                'Authorization': 'Basic ' + this.currentApiKey,
            },
        });
        if (res.status != 200) {
            throw `createShare failed. status: ${res.status} url: ${url}`;
        }
    }

    async getBuildTargetDetail(buildTargetId: string) : Promise<BuildTargetDetailInfo> {
        this.checkApiKey();
        const orgAndProject = this.getOrgAndProjectParams();
        const client = await this.api.init<CloudBuildClient>();
        const res = await client.getBuildTarget({...orgAndProject, buildtargetid: buildTargetId})
        console.log('getBuildTargetDetail result:', res.data);
        const data = res.data as GetBuildTargetResponseItemType;
        if (data == null) {
            throw "getBuildTargetDetail failed.";
        }
        return new BuildTargetDetailInfo(data);
    }
}

export class ProjectInfo
{
    name: string;
    projectId: string;
    orgName: string;
    orgId: string;
    projectDetail: any;

    constructor(project: GetListProjectsForUserResponseItemType) {
        this.name = project.name;
        this.projectId = project.projectid;
        this.orgName = project.orgName;
        this.orgId = project.orgid;
        this.projectDetail = project;
    }
}

// simple data model
export class BuildInfo
{
    build?: number;
    buildTargetId?: string;
    buildTargetName?: string;
    logUrl?: string;
    downloadUrl?: string;
    buildStatus: string;
    commitId?: string;
    scmBranch?: string;
    detailText: string;
    totalTimeInSeconds?: number;
    buildStartTime?: Date;
    checkoutStartTime?: Date;
    finishedTime?: Date;
    createdTime?: Date;
    createShareUrl?: string;

    constructor(build: GetBuildsResponseItemType) {
        console.log("creating BuildInfo.", build);
        this.build = build.build;
        this.buildTargetId = build.buildtargetid;
        this.buildTargetName = build.buildTargetName;
        const logLink = build.links?.log?.href;
        this.downloadUrl = build.links?.download_primary?.href ?? "";
        this.logUrl = (logLink != null) ? (cloudBuildLogUrlBase + logLink) : "";
        this.buildStatus = build.buildStatus;
        this.commitId = build.lastBuiltRevision;
        this.scmBranch = build.scmBranch;
        this.detailText = JSON.stringify(build, null, 2);
        this.totalTimeInSeconds = build.totalTimeInSeconds;
        this.buildStartTime = (build.buildStartTime != null) ? new Date(build.buildStartTime) : null;
        this.checkoutStartTime = (build.checkoutStartTime != null) ? new Date(build.checkoutStartTime) : null;
        this.finishedTime = (build.finished != null) ? new Date(build.finished) : null;
        this.createdTime = (build.created != null) ? new Date(build.created) : null;
        const createShareLink = build.links?.create_share?.href;
        this.createShareUrl = (createShareLink != null) ? (cloudBuildLogUrlBase + createShareLink) : null;
    }

    getLogTextUri(): vscode.Uri {
        const uri = vscode.Uri.parse(cloudBuildLogScheme + ":" + this.logUrl + "/" + this.buildTargetId + "-" + this.build);
        return uri;
    }

    getCurrentBuildSec(): number {
        if (this.buildStatus != "started") {
            return null;
        }
        if (this.checkoutStartTime == null) {
            return null;
        }
        var sec = Math.floor((Date.now() - this.checkoutStartTime.getTime()) / 1000);
        return sec;
    }
}

export class BuildTargetInfo
{
    buildTargetId: string;
    name: string;
    platform?: string;
    detailText: string;

    constructor(buildtarget: GetBuildTargetsResponseItemType) {
        this.buildTargetId = buildtarget.buildtargetid;
        this.name = buildtarget.name;
        this.platform = buildtarget.platform;
        this.detailText = JSON.stringify(buildtarget, null, 2);
    }
}

export class BuildTargetDetailInfo
{
    buildTargetId: string;
    name: string;
    platform?: string;
    detailText: string;

    constructor(buildtarget: GetBuildTargetResponseItemType) {
        this.buildTargetId = buildtarget.buildtargetid;
        this.name = buildtarget.name;
        this.platform = buildtarget.platform;
        this.detailText = JSON.stringify(buildtarget, null, 2);
    }
}

export class ShareInfo
{
    shareId?: string;
    shareExpirationTime: Date;

    constructor(share: GetShareResponseItemType) {
        this.shareId = share.shareid;
        this.shareExpirationTime = (share.shareExpiry != null) ? new Date(share.shareExpiry) : null;
    }
}

src/extension.ts
import * as vscode from 'vscode';
import { CloudBuildLogContentProvider } from './logprovider';
import { BuildTreeDataProvider, BuildTreeItem, HeaderTreeItem } from './treeprovider';
import { ApiLoader, BuildTargetInfo, ProjectInfo } from './apiloader';
import { BuildDetailContentProvider } from './builddetailprovider';

const cloudBuildLogScheme = "unitycloudbuildviewer";
const cloudBuildDetailScheme = "unitycloudbuilddetail"
//const cloudBuildLogUrlBase = "https://build-api.cloud.unity3d.com";

var treeRedrawTimer: NodeJS.Timer = null;

export function activate(context: vscode.ExtensionContext) {
	console.log('Congratulations, your extension "unitycloudbuild-viewer" is now active!');

	// setup Build Log Virtual Document
	const logProvider = new CloudBuildLogContentProvider(() => getApiKey());
	context.subscriptions.push(vscode.workspace.registerTextDocumentContentProvider(cloudBuildLogScheme, logProvider));

	// API loader
	const apiLoader = new ApiLoader(() => getApiKey(), context, context.extensionPath + "/cloudbuildapi.json");

	// setup Build Detail Virtual Document
	const buildDetailProvider = new BuildDetailContentProvider(apiLoader);
	context.subscriptions.push(vscode.workspace.registerTextDocumentContentProvider(cloudBuildDetailScheme, buildDetailProvider));

	// setup Tree View
	const buildTreeDataProvider = new BuildTreeDataProvider(apiLoader);
	const treeView = createTreeView(buildTreeDataProvider, buildDetailProvider);

	// load Build Configs (speedup filter selection)
	if (apiLoader.isConfigured()) {
		apiLoader.getBuildTargetsCached(false);
	}

	// redraw tree view (update time information)
	treeRedrawTimer = setInterval(() => {
		console.log("interval redraw");
		buildTreeDataProvider.redraw();
	}, 30000);

	context.subscriptions.push(vscode.commands.registerCommand('cloudbuildexplorer.refreshEntry', () =>
		buildTreeDataProvider.reload()
	));

	context.subscriptions.push(vscode.commands.registerCommand('cloudbuildexplorer.clearFilter', () => {
		buildTreeDataProvider.resetFilter();
		buildTreeDataProvider.reload();
	}));


	context.subscriptions.push(vscode.commands.registerCommand('cloudbuildexplorer.filterBuilds', async () => {
		processBuildFilter(apiLoader, buildTreeDataProvider);
	}));

	context.subscriptions.push(vscode.commands.registerCommand('cloudbuildexplorer.viewTextLog', async (build: BuildTreeItem) => {
		if (build.buildInfo.logUrl != null) {
			let uri = build.buildInfo.getLogTextUri();
			let doc = await vscode.workspace.openTextDocument(uri);
			await vscode.window.showTextDocument(doc, { preview: false });
		}
	}));

	context.subscriptions.push(vscode.commands.registerCommand('cloudbuildexplorer.buildAgain', async (build: BuildTreeItem) => {
		try {
			var result = await apiLoader.startBuild(build.buildInfo.buildTargetId);
			if (result.error == null) {
				buildTreeDataProvider.reload();
			}
			else
			{
				vscode.window.showErrorMessage(`Failed building '${build.buildInfo.buildTargetId}'. error: ${result.error}`);
			}
		}
		catch(e) {
			vscode.window.showErrorMessage(`Failed building '${build.buildInfo.buildTargetId}'. error: ${e}`);
		}
	}));

	context.subscriptions.push(vscode.commands.registerCommand('cloudbuildexplorer.cancelBuild', async (build: BuildTreeItem) => {
		try {
			var result = await apiLoader.cancelBuild(build.buildInfo.buildTargetId, build.buildInfo.build);
			buildTreeDataProvider.reload();
		}
		catch(e) {
			vscode.window.showErrorMessage(`Failed canceling '${build.buildInfo.buildTargetId} #${build.buildInfo.build}'. error: ${e}`);
		}
	}));

	context.subscriptions.push(vscode.commands.registerCommand('cloudbuildexplorer.createShare', async (build: BuildTreeItem) => {
		if (build.buildInfo.buildStatus != "success") {
			vscode.window.showErrorMessage(`Create Share Link is available for successful builds only.`);
			return;
		}
		try {
			console.log("createShareUrl:" + build.buildInfo.createShareUrl);
			await apiLoader.createShare(build.buildInfo.createShareUrl);
			buildTreeDataProvider.redraw();
		}
		catch(e) {
			vscode.window.showErrorMessage(`Failed building '${build.buildInfo.buildTargetId}'. error: ${e}`);
		}
	}));

	context.subscriptions.push(vscode.commands.registerCommand('unitycloudbuild-viewer.setup', async () => {
		if (getApiKey() == "") {
			var selection = await vscode.window.showErrorMessage("Before setup, please set the Api Key setting in the User Settings.", "Open Settings");
			if (selection == "Open Settings") {
				openSettings();
			}
			return;
		}
		// setup
		let projects: ProjectInfo[];
		try {
			projects = await apiLoader.readProjects();
		}
		catch(e) {
			console.error(e);
			vscode.window.showErrorMessage('Setup of UnityCloudBuildViewer failed! ApiKey setting may be wrong.');
			return;
		}
		const pickItems = projects.map(p => new (class implements vscode.QuickPickItem {
			label: string = `"${p.name}" project of "${p.orgName}" organization`;
			project: ProjectInfo = p;
		}));
		const result = await vscode.window.showQuickPick(pickItems, {
			placeHolder: 'Please select the target project',
		})
		if (result == null) {
			vscode.window.showInformationMessage('Setup of UnityCloudBuildViewer canceled.');
			return;
		}
		await context.globalState.update("projectId", result.project.projectId);
		await context.globalState.update("orgId", result.project.orgId);
		buildTreeDataProvider.reload();
		apiLoader.getBuildTargetsCached(false);
		//vscode.commands.executeCommand("onView:cloudbuildexplorer");
		vscode.window.showInformationMessage('Setup of UnityCloudBuildViewer completed!');
	}));

	context.subscriptions.push(vscode.commands.registerCommand('unitycloudbuild-viewer.reload', () => {
		if (!vscode.window.activeTextEditor) {
			return;
		  }
		  let { document } = vscode.window.activeTextEditor;
		  if (document.uri.scheme !== cloudBuildLogScheme) {
			return;
		  }
		  logProvider.reload(document.uri);
	}));

	context.subscriptions.push(vscode.commands.registerCommand('unitycloudbuilddetail.viewTextLog', async () => {
		let uri = buildDetailProvider.currentBuild.getLogTextUri();
		let doc = await vscode.workspace.openTextDocument(uri);
		await vscode.window.showTextDocument(doc, { preview: false });
	}));
}

// this method is called when your extension is deactivated
export function deactivate() {
	clearInterval(treeRedrawTimer);
	treeRedrawTimer = null;
}

function getApiKey() : string
{
	return vscode.workspace.getConfiguration('unitycloudbuild-viewer')?.get("apiKey") ?? "";
}

function createTreeView(buildTreeDataProvider: BuildTreeDataProvider, buildDetailProvider: BuildDetailContentProvider): vscode.TreeView<BuildTreeItem | HeaderTreeItem> {
	const treeView = vscode.window.createTreeView(
		'cloudbuildexplorer',
		{treeDataProvider: buildTreeDataProvider}
	);
	treeView.onDidChangeSelection( async (e) => {
		const treeItem = e.selection[0];
		if (!(treeItem instanceof BuildTreeItem)) {
			return;
		}
		const buildTreeItem = treeItem as BuildTreeItem;
		buildDetailProvider.setCurrentBuild(buildTreeItem.buildInfo);
		let uri = vscode.Uri.parse(cloudBuildDetailScheme + ":/Cloud Build Detail");
		let doc = await vscode.workspace.openTextDocument(uri);
		await vscode.window.showTextDocument(doc, { preview: false });
	});
	return treeView;
}

function openSettings()
{
	vscode.commands.executeCommand('workbench.action.openSettings', "Unity Cloud Build Viewer");
}

// show configs selector and apply filter to tree view
async function processBuildFilter(apiLoader: ApiLoader, buildTreeDataProvider: BuildTreeDataProvider)
{
	class BuildTargetPickItem implements vscode.QuickPickItem {
		label: string;
		buildTarget: BuildTargetInfo;
		constructor(buildTarget: BuildTargetInfo) {
			this.label = buildTarget.name;
			this.buildTarget = buildTarget;
		}
	}

	class ReloadPickItem implements vscode.QuickPickItem {
		label: string;
		constructor() {
			this.label = "Reload Configs...";
		}
	}

	const buildTargets = await getBuildTargetsWithProgress(apiLoader, false);
	const sortedBuildTargets = buildTargets.sort((a,b) =>  (a.name > b.name ? 1 : -1));
	var pickItems : (BuildTargetPickItem|ReloadPickItem)[] = sortedBuildTargets.map(i => new BuildTargetPickItem(i));
	pickItems.unshift(new ReloadPickItem());
	const options = new class implements vscode.QuickPickOptions {
		canPickMany: boolean;
		title: string;
		placeHolder: string;
		constructor() {
			this.canPickMany = false;
			this.title = "Please select a config to filter builds.";
			this.placeHolder = "Filter by config name...";
		}
	};
	const selected = await vscode.window.showQuickPick(pickItems, options);
	if (selected == null) {
		return;
	}
	if (selected instanceof ReloadPickItem) {
		await getBuildTargetsWithProgress(apiLoader, true);
		vscode.commands.executeCommand('cloudbuildexplorer.filterBuilds');
		return;
	}
	buildTreeDataProvider.filterByConfig((selected as BuildTargetPickItem).buildTarget);
	buildTreeDataProvider.reload();
	vscode.window.showInformationMessage('picked filter:' + selected.label);
}

async function getBuildTargetsWithProgress(apiLoader: ApiLoader, reload: boolean): Promise<BuildTargetInfo[]>
{
	return await vscode.window.withProgress<BuildTargetInfo[]>({
		location: vscode.ProgressLocation.Notification,
		cancellable: false,
		title: 'Loading CloudBuild configs...'
	}, async (progress) => {
		let buildTargetInfo: BuildTargetInfo[];
		progress.report({ increment: 0 });
		buildTargetInfo = await apiLoader.getBuildTargetsCached(reload);
		progress.report({ increment: 100 });
		return buildTargetInfo;
	});
}


src/builddetailprovider.ts
import { createCipheriv } from 'crypto';
import * as vscode from 'vscode';
import { ApiLoader, BuildInfo, BuildTargetDetailInfo } from './apiloader';

const shareLinkUrlBase = "https://developer.cloud.unity3d.com/share/share.html?shareId=";

export class BuildDetailContentProvider implements vscode.TextDocumentContentProvider {
	uri: vscode.Uri;
    currentBuild: BuildInfo;
	onDidChangeEmitter = new vscode.EventEmitter<vscode.Uri>();
	onDidChange = this.onDidChangeEmitter.event;
	apiLoader: ApiLoader;
	shareLink: string;
	buildTargetDetail: BuildTargetDetailInfo;

	constructor(apiLoader: ApiLoader) {
		this.apiLoader = apiLoader;
		console.log("CloudBuildLogContentProvider constructor");
	}

	getDurationStringFromSec(sec : number) : string {
		if (sec == null) {
			return "";
		}
		return `${Math.floor(sec / 60)}min ${Math.floor(sec) % 60}sec`;
	}

	provideTextDocumentContent(uri: vscode.Uri): string {
		console.warn("provideTextDocumentContent called.");
		this.uri = uri;
		return "Name: " + "`" + this.getBuildName() + "`" + "\n" + 
		"Status: " + this.currentBuild.buildStatus + "\n" +
		"Download: " + this.currentBuild.downloadUrl + "\n" +
		"Branch: " + this.currentBuild.scmBranch + "\n" +
		"Commit Id: " + this.currentBuild.commitId + "\n" +
		this.buildTimeText() + "\n" +
		this.shareLinkText() + "\n" +
		"\n" +
		"Detail: " + this.currentBuild.detailText + "\n" +
		"\n" +
		this.configText() + "\n";
	}

	buildTimeText() : string {
		if (this.currentBuild.buildStatus == "started") {
			return "Building For: " + this.getDurationStringFromSec(this.currentBuild.getCurrentBuildSec());
		}
		return "Build Time: " + this.getDurationStringFromSec(this.currentBuild.totalTimeInSeconds);
	}

	shareLinkText() : string {
		if (this.shareLink === undefined) {
			return "Share Link: (loading...)";
		}
		if (this.shareLink === null) {
			return "Share Link: ";
		}
		return "Share Link: " + "`" + this.getBuildName() + "` " + shareLinkUrlBase + this.shareLink;
	}

	configText() : string {
		if (this.buildTargetDetail === undefined) {
			return "Config: (loading...)";
		}
		if (this.buildTargetDetail === null) {
			return "Config: (not available)";
		}
		return "Config: " + this.buildTargetDetail?.detailText;
	}

	getBuildName() {
		return this.currentBuild.buildTargetId + " #" + this.currentBuild.build;
	}

	setCurrentBuild(build : BuildInfo) {
		this.currentBuild = build;
		this.shareLink = undefined; // undefined means loading. null means no data or error.
		this.buildTargetDetail = undefined; // undefined means loading. null means no data or error.
		this.ShareLinkGetter(build); // async method but don't await
		this.BuildTargetDetailGetter(build);
		this.onDidChangeEmitter.fire(this.uri);
	}

	async ShareLinkGetter(build : BuildInfo) {
		// share link is not needed for failed or running builds
		if (build.buildStatus != "success") {
			console.warn("set shareLink = null");
			this.shareLink = null;
			this.onDidChangeEmitter.fire(this.uri);
			return;
		}
		try {
			var share = await this.apiLoader.getShare(build.buildTargetId, build.build);
		}
		catch(e) {
			console.log("ShareLinkGetter error: ", e);
			share = null;
			// nullの結果を表示するため、このまま下に流す
		}
		// APIのレスポンスを待っている間に表示対象ビルドが変わっていることがあるので、その場合は捨てる。
		if (this.currentBuild.buildTargetId != build.buildTargetId || this.currentBuild.build != build.build) {
			return;
		}
		this.shareLink = share?.shareId;
		this.onDidChangeEmitter.fire(this.uri);
	}

	async BuildTargetDetailGetter(build : BuildInfo) {
		try {
			var buildTargetDetail = await this.apiLoader.getBuildTargetDetail(build.buildTargetId);
		}
		catch(e) {
			console.log("BuildTargetDetailGetter error: ", e);
			buildTargetDetail = null;
			// nullの結果を表示するため、このまま下に流す
		}
		// APIのレスポンスを待っている間に表示対象ビルドが変わっていることがあるので、その場合は捨てる。
		if (this.currentBuild.buildTargetId != build.buildTargetId || this.currentBuild.build != build.build) {
			return;
		}
		this.buildTargetDetail = buildTargetDetail
		this.onDidChangeEmitter.fire(this.uri);
	}
}


src/treeprovider.ts
import { hasUncaughtExceptionCaptureCallback } from 'process';
import * as vscode from 'vscode';
import { ApiLoader, BuildInfo, BuildTargetInfo } from './apiloader';

type TreeItem = BuildTreeItem | HeaderTreeItem;

export class BuildTreeDataProvider implements vscode.TreeDataProvider<TreeItem> {
    apiLoader: ApiLoader;
    private onDidChangeTreeDataEmitter: vscode.EventEmitter<BuildTreeItem | undefined | null | void> = new vscode.EventEmitter<BuildTreeItem | undefined | null | void>();
    readonly onDidChangeTreeData: vscode.Event<BuildTreeItem | undefined | null | void> = this.onDidChangeTreeDataEmitter.event;
    buildTargetFilter?: BuildTargetInfo;
    treeItems?: TreeItem[];

    itemsPerPage: number = 20;
    maxPages: number = 10;
    loadId: number = 0;

    constructor(apiLoader: ApiLoader) {
        this.apiLoader = apiLoader;
    }

    // load again from server and redraw
    reload(): void {
        this.treeItems = null;
        this.onDidChangeTreeDataEmitter.fire();
    }

    // draw again without changing contents
    redraw(): void {
        this.onDidChangeTreeDataEmitter.fire();
    }

    filterByConfig(buildTarget?: BuildTargetInfo): void
    {
        this.buildTargetFilter = buildTarget;
    }

    buildTargetIdForFilter(): string
    {
        if (this.buildTargetFilter == null) {
            return null;
        }
        return this.buildTargetFilter.buildTargetId;
    }

    currentFilterText(): string
    {
        if (this.buildTargetFilter == null) {
            return null;
        }
        return this.buildTargetFilter.name;
    }

    resetFilter(): void {
        this.buildTargetFilter = null;
    }

    getTreeItem(element: TreeItem): vscode.TreeItem {
        return element;
    }

    getChildren(element?: TreeItem): Thenable<TreeItem[]> {
        if (element) {
            return Promise.resolve([]);
        } else {
            return this.getRootItems();
        }
    }

    getRootItems(): Thenable<TreeItem[]>
    { 
        if (!this.apiLoader.isConfigured()) {
            return Promise.resolve([]);
        }
        return new Promise((resolve, reject) => {
            this.FetchRootItems(resolve, reject);
        });
    }

    // ビルドアイテムを取得する非同期メソッド。
    // ロード済みの場合はそのまま返す。
    // コンテンツ未ロードの場合はサーバーから取得して１ページ目を返す。
    // １ページ目をロードしたとき、２ページ目以降のロードも仕掛けて、ロードされ次第再描画する。
    async FetchRootItems(resolve: (value: TreeItem[]) => void, reject) {
        if (this.treeItems != null) {
            // update description for time display
            this.treeItems.forEach(i => {
                if (i instanceof BuildTreeItem) {
                    i.updateForCurrentTime();
                }
            });
            resolve(this.treeItems);
            return;
        }
        try {
            const onPage: ((loadId: number, page: number, builds: TreeItem[]) => void) = (loadId, page, builds) => {
                // 古い指示に対する結果は不要のため破棄する
                if (loadId != this.loadId) {
                    return;
                }
                if (this.treeItems != null) {
                    this.treeItems.push(...builds);
                    this.redraw();
                }
            };
            this.loadId++;
            if (this.buildTargetIdForFilter() != null) {
                var builds = await this.FetchFilteredBuildItems();
                this.FetchBuildPagesAsync(this.loadId, this.buildTargetIdForFilter(), 2, this.itemsPerPage, onPage);
                this.treeItems = builds;
                resolve(builds);
            }
            else {
                var builds = await this.FetchAllBuildItems(this.loadId, this.itemsPerPage);
                // 値の返却後も次ページをロードする（完了を待たない。onPageが順次呼ばれる）
                this.FetchBuildPagesAsync(this.loadId, null, 2, this.itemsPerPage, onPage);
                this.treeItems = builds;
                resolve(builds);
            }
        }
        catch(e) {
            reject(e);
        }
    }

    // fetch all build items
    // 全ビルドの１ページ目取得に加え、実行中のビルド (buildStatus = started) も平行して取得し、必ず表示する。
    // 長時間実行中のビルドが１ページ目に入らないことがあることの対策。
    async FetchAllBuildItems(loadId: number, itemsPerPage: number)
        : Promise<TreeItem[]>
    {
        const [allBuilds, startedBuilds] = await Promise.all([this.apiLoader.getBuilds(null, null, 1, itemsPerPage), this.apiLoader.getBuilds(null, "started")]);
        if (allBuilds == null || startedBuilds == null) {
            throw new Error("getBuilds return null.");
        }
        if (allBuilds.length == 0 && startedBuilds.length == 0) {
            return [new HeaderTreeItem("No builds found", null)];
        }
        const startedBuildItems = startedBuilds.map(i => new BuildTreeItem(i));
        // startedのものは除外することで二重表示しないようにする
        const allBuildItems = allBuilds.map(i => (i.buildStatus != "started") ? new BuildTreeItem(i) : null);

        return [...startedBuildItems, ...allBuildItems];
    }

    async FetchFilteredBuildItems() : Promise<TreeItem[]> {
        const builds = await this.apiLoader.getBuilds(this.buildTargetIdForFilter());
        if (builds != null) {
            if (builds.length > 0) {
                const buildItems = builds.map(i => new BuildTreeItem(i));
                // 結果がフィルタされている場合はヘッダ表示を追加
                return [new HeaderTreeItem(`<< Filter: ${this.currentFilterText()} >>`, "filter"), ...buildItems];
            }
            else {
                return [new HeaderTreeItem(`<< Filter: ${this.currentFilterText()} >>`, "filter"), new HeaderTreeItem("No builds found", null)];
            }
        }
        else {
            throw new Error("getBuilds return null.");
        }
    }

    // ２ページ目以降をロードする
    // ロードが進むたびにonPageが呼ばれる
    async FetchBuildPagesAsync(loadId: number, buildTargetId: string, page: number, itemsPerPage: number, onPage: (loadId: number, page: number, builds: TreeItem[]) => void) {
        for (var i=2 ; i<=this.maxPages ; i++) {
            const builds = await this.FetchBuildPageAsync(i, buildTargetId, itemsPerPage);
            onPage(loadId, page, builds);
            if (builds.length == 0) {
                break;
            }
        }
    }

    // 次ページ取得
    async FetchBuildPageAsync(page: number, buildTargetId: string, itemsPerPage: number) : Promise<TreeItem[]> {
        const allBuilds = await this.apiLoader.getBuilds(buildTargetId, null, page, itemsPerPage);
        if (allBuilds == null) {
            throw new Error("getBuilds return null.");
        }
        if (allBuilds.length == 0) {
            return [];
        }
        // startedのものは除外することで二重表示しないようにする
        const allBuildItems = allBuilds.map(i => (i.buildStatus != "started") ? new BuildTreeItem(i) : null);

        return allBuildItems;
    }
}

// custom tree item for builds
export class BuildTreeItem extends vscode.TreeItem {
    logUrl: string;
    buildInfo: BuildInfo;
    constructor(
        buildInfo: BuildInfo
    ) {
        super(buildInfo.buildTargetName + " #" + buildInfo.build, vscode.TreeItemCollapsibleState.None);
        this.buildInfo = buildInfo;
        this.contextValue = "builditem";
        this.updateForCurrentTime();
    }

    getBuildStatusAndDuration() : string {
        if (this.buildInfo.buildStatus == "started") {
            return "building for " + this.getDurationStringFromDate(this.getLastChangeDate());
        }
        return this.buildInfo.buildStatus + " " + this.getDurationStringFromDate(this.getLastChangeDate()) + " ago";
    }

	getDurationStringFromDate(date : Date) : string {
		if (date == null) {
			return "";
		}
        var min = Math.floor((Date.now() - date.getTime()) / (1000 * 60));
        if (min < 60) {
            return `${min}m`;
        }
        if (min < 60 * 24 * 2) {
            return `${(min/60).toFixed()}h`;
        }
        return `${(min/(60 * 24)).toFixed()}d`;
	}

    getLastChangeDate(): Date {
        switch (this.buildInfo.buildStatus) {
            case "started":
                return this.buildInfo.checkoutStartTime;
            case "queued":
            case "sentToBuilder":
                return this.buildInfo.createdTime;
        }
        return this.buildInfo.finishedTime;
    }

    updateForCurrentTime() {
        this.description = this.getBuildStatusAndDuration();
    }
}

// no item for builds
export class HeaderTreeItem extends vscode.TreeItem {
    constructor(label: string, contextValue: string
    ) {
        super(label, vscode.TreeItemCollapsibleState.None);
        this.contextValue = contextValue;
    }
}
